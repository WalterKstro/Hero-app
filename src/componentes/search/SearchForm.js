import useForm from '../../hooks/useForm'
import queryString  from 'query-string'
import { useLocation } from 'react-router-dom'
import HeroCard from '../heroes/HeroCard'
import getByName from '../../switcher/getByName'
import { useMemo } from 'react'


const SearchForm = ({history}) => {
    
    const {search} = useLocation()
    const {q=''} = queryString.parse(search);
    const {state,handleChangeInput} = useForm()
    
    const handleSearch = e => {
        e.preventDefault()
        if (state.hasOwnProperty('search')) {
            history.push(`?q=${state.search}`)
        }
    }

    const herosFilteredById = useMemo(() => getByName(q), [q])

    return (
        <>
            <div className="row my-5">
                <div className="col-md-4 offset-md-4">
                <form>
                    <div className="mb-3">
                        <label htmlFor="heroe" className="form-label">Name heroe</label>
                        <input 
                            type="text" 
                            name="search"
                            className="form-control" 
                            id="heroe"
                            value={state.search}
                            onChange = { handleChangeInput }
                        />
                    </div>
                    <button 
                        type="submit" 
                        className="btn btn-outline-primary"
                        onClick={handleSearch}
                    >Search heroe</button>
                </form>
                </div>
            </div>
            <div className="row">
                <div className="col-12 wrapper_grid">
                    {
                        herosFilteredById.length < 1
                        ?
                            <div className="alert alert-primary alert-wrapper" role="alert">
                                ðŸ˜ž Not found heroes with the name  <strong>{q}</strong>
                            </div>
                        :
                            herosFilteredById.map(heroe => 
                                <HeroCard
                                    key={heroe.id}
                                    id={heroe.id}
                                    superhero={heroe.superhero}
                            />
                        )
                    }
                </div>
            </div>
        </>
    )
}

export default SearchForm
