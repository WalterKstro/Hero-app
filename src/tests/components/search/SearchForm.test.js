import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import SearchForm from "../../../componentes/search/SearchForm"

describe('Testing in <SearchForm/>', () => {

    test('should display correctly', () => {

        const rendered = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchForm} />
            </MemoryRouter>
        )

        expect(rendered).toMatchSnapshot()

    })


    test('should display a result of queryString', () => {
        const rendered = mount(
            <MemoryRouter initialEntries={['/search?q=Superman']}>
                <Route path="/search" component={SearchForm} />
            </MemoryRouter>

        )

        expect(rendered.find('HeroCard').exists()).toBe(true)
        expect(rendered).toMatchSnapshot()
    })

    test('should display an alert if not found a heroe with the queryString', () => {
        const rendered = mount(
            <MemoryRouter initialEntries={['/search?q=superman12']}>
                <Route path="/search" component={SearchForm} />
            </MemoryRouter>
        )
        expect(rendered.find('.alert').exists()).toBe(true)
        expect(rendered).toMatchSnapshot()

    })
    
    
    
})
