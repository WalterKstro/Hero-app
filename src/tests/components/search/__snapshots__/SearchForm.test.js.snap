// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <SearchForm/> should display a result of queryString 1`] = `
Array [
  <div
    className="row my-5"
  >
    <div
      className="col-md-4 offset-md-4"
    >
      <form>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="heroe"
          >
            Name heroe
          </label>
          <input
            className="form-control"
            id="heroe"
            name="search"
            onChange={[Function]}
            type="text"
          />
        </div>
        <button
          className="btn btn-outline-primary"
          onClick={[Function]}
          type="submit"
        >
          Search heroe
        </button>
      </form>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-12 wrapper_grid"
    >
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-superman.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Superman
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-superman"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Testing in <SearchForm/> should display an alert if not found a heroe with the queryString 1`] = `
Array [
  <div
    className="row my-5"
  >
    <div
      className="col-md-4 offset-md-4"
    >
      <form>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="heroe"
          >
            Name heroe
          </label>
          <input
            className="form-control"
            id="heroe"
            name="search"
            onChange={[Function]}
            type="text"
          />
        </div>
        <button
          className="btn btn-outline-primary"
          onClick={[Function]}
          type="submit"
        >
          Search heroe
        </button>
      </form>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-12 wrapper_grid"
    >
      <div
        className="alert alert-primary alert-wrapper"
        role="alert"
      >
        ðŸ˜ž Not found heroes with the name  
        <strong>
          superman12
        </strong>
      </div>
    </div>
  </div>,
]
`;

exports[`Testing in <SearchForm/> should display correctly 1`] = `
Array [
  <div
    className="row my-5"
  >
    <div
      className="col-md-4 offset-md-4"
    >
      <form>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="heroe"
          >
            Name heroe
          </label>
          <input
            className="form-control"
            id="heroe"
            name="search"
            onChange={[Function]}
            type="text"
          />
        </div>
        <button
          className="btn btn-outline-primary"
          onClick={[Function]}
          type="submit"
        >
          Search heroe
        </button>
      </form>
    </div>
  </div>,
  <div
    className="row"
  >
    <div
      className="col-12 wrapper_grid"
    >
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-batman.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Batman
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-batman"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-superman.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Superman
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-superman"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-flash.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Flash
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-flash"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-green.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Green Lantern
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-green"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-arrow.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Green Arrow
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-arrow"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-wonder.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Wonder Woman
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-wonder"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-martian.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Martian Manhunter
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-martian"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-robin.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Robin/Nightwing
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-robin"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-blue.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Blue Beetle
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-blue"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/dc-black.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Black Canary
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/dc-black"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-spider.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Spider Man
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-spider"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-captain.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Captain America
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-captain"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-iron.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Iron Man
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-iron"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-thor.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Thor
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-thor"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-hulk.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Hulk
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-hulk"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-wolverine.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Wolverine
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-wolverine"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-daredevil.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Daredevil
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-daredevil"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-hawkeye.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Hawkeye
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-hawkeye"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-cyclops.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Cyclops
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-cyclops"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
      <div
        className="card animate__animated animate__zoomIn cover_hero"
      >
        <img
          alt="..."
          className="img-fluid card-img-top"
          src="/assets/heroes/marvel-silver.jpg"
        />
        <div
          className="card-body cover__overflow"
        >
          <h5
            className="card-title text-white bolder"
          >
            Silver Surfer
          </h5>
          <a
            className="btn btn-light text-bolder"
            href="/hero/marvel-silver"
            onClick={[Function]}
          >
            View More
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;
